operator/pkg/gateway-api/grpcroute.go

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package gateway_api
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	gatewayv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/cilium/cilium/operator/pkg/gateway-api/helpers"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// grpcRouteReconciler reconciles a GRPCRoute object
+type grpcRouteReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	Model *internalModel
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *grpcRouteReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &gatewayv1alpha2.GRPCRoute{}, backendServiceIndex,
+		func(rawObj client.Object) []string {
+			hr, ok := rawObj.(*gatewayv1alpha2.GRPCRoute)
+			if !ok {
+				return nil
+			}
+			var backendServices []string
+			for _, rule := range hr.Spec.Rules {
+				for _, backend := range rule.BackendRefs {
+					if !helpers.IsService(backend.BackendObjectReference) {
+						continue
+					}
+					backendServices = append(backendServices,
+						types.NamespacedName{
+							Namespace: helpers.NamespaceDerefOr(backend.Namespace, hr.Namespace),
+							Name:      string(backend.Name),
+						}.String(),
+					)
+				}
+			}
+			return backendServices
+		},
+	); err != nil {
+		return err
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &gatewayv1alpha2.GRPCRoute{}, gatewayIndex,
```

Comment:
This line is is some proper arcane incantation. Can you please add a comment :-)?